### ISSUE TITLE
Build warp sync based state sync as a plugin

### ISSUE THREAD

**Issue by cburgdorf** [ethereum/py-evm#1593]
### What is wrong?

One of the most challenging problems for all of the existing Ethereum 1 client implementations today is how to keep the state sync fast.

The reason for that is not so much the pure size of the state (currently ~9GB) but rather the number of trie nodes which is somewhere above 200 million. As far as I understand there are various reasons why syncing these is hard:

1. The networking protocol that is being used does only allow up to `n` (need to lookup)  trie nodes to be fetched per request

2. Client nodes that aren't archive nodes prune past state after `n` bocks which means there is only a limited amount of time until it becomes very hard to find the trie nodes on the network.

### How can it be fixed

A while back parity introduces *warp sync* sync to speed up the entire syncing process.  For the sake of simplicity, I'll focus on how to use warp sync to just sync the state and why this would not compromise security.

Warp sync works by taking a snapshot of the entire chain + state every 5000 blocks and making that available to other peers through an extension of the existing wire protocol.

The snapshot is divided into a *snapshot manifest* and the actual *snapshot data*. All of that is RLP encoded data.

**Manifest**

```
[
    version: P, // snapshot format version. Must be set to 2.
    state_hashes: [hash_1: B_32, hash_2: B_32, ...], // a list of all the state chunks in this snapshot
    block_hashes: [hash_1: B_32, hash_2: B_32, ...], // a list of all the block chunks in this snapshot
    state_root: B_32, // the root which the rebuilt state trie should have. used to ensure validity
    block_number: P, // the number of the best block in the snapshot; the one which the state coordinates to.
    block_hash: B_32, // the best block in the snapshot's hash.
]
```   

As we can see, the manifest contains the `state_root` as well as `state_hashes` for state chunks of 4 MB.

What we can do is to request the state for a given `state_root` by fetching these 4 MB chunks and then ensuring integrity by recomputing the `state_root` from the entire state once we've fetched all pieces.

In terms of security this would be pretty identical to the existing state sync and the worst thing that could happen is that we can only fully ensure the integrity once we've finished downloading the entire state.

Or to say cite @AlexeyAkhunov on the topic

>Warp sync is pretty much equivalent to fast sync in terms of trust required. What you do when you warp sync, is downloading these (7Gb?) data, then compute Patricia Merkle tree over this, and check that the state root is the same as in the block header. Done. With the fast sync (current geth's method), you build the state trie from the nodes you receive over the network, and in the end you are also sure that you have the state matching the state root in the block header. The main difference is that when you warp sync, you have to download ALL 7Gb before you are sure you are actually downloading the state, and not some random data, wheres with the fast sync, you pretty much know that you are downloading correct stuff as you do it. If the fast sync peer sends you rubbish, you will realise very quickly without having to download gigabytes

I'm not proposing to make this the default state sync strategy (as it is non-standard for the time being) but I *think* this could be useful as a plugin to give a user an alternative option for a fast(er) state sync.

**Comment by AlexeyAkhunov** [comment]
There is a way of combining advantages of warp sync and fast sync. The rough idea is this. First, you flatten the state as list of pairs (account;account_data) sorted by the hash of the account address (this is how it appears in the Patricia tree). Then, you chunk it up, lets say, in 4Mb each chunk. Each chunk will contain pairs that stand next to each other in the patricia tree. What you add to the chunk is a proof that this chunk is indeed part of the state with the given root. Once you have got all the account, you repeat this procedure for the contract storage. For contracts with small storage, you don't bother making proofs, but just package them into chunks. But for contracts with the large storage, you use the same technique described above.

I almost have a code for this in Turbo-Geth! :) Perhaps I shall try to code it up and give it a go!

**Comment by cburgdorf** [comment]
>Each chunk will contain pairs that stand next to each other in the patricia tree. What you add to the chunk is a proof that this chunk is indeed part of the state with the given root.

Am I write to assume that in order to verify the chunks I need to have them in order? So, if I fetch lots of chunks in parallel, and I have gaps, then I need to fill out the gaps before I can verify those because the proofs do not work individually but only in conjunction with the neighbor. Or am I totally off with that?

>I almost have a code for this in Turbo-Geth! :) Perhaps I shall try to code it up and give it a go!

Sounds very exciting! :+1: 

But, since this would be a totally new syncing scheme, there are no nodes to seed this. 
Where my initial intend was to create a state sync that can benefit from the existing parity nodes that provide warp sync chunks, this would be something entirely new that needs to be adopted (or am I misinterpreting this?)

But if your proposed scheme does indeed work and improve the speed of the state sync, I guess it would be worth drafting an EIP for that :)

**Comment by AlexeyAkhunov** [comment]
> Am I write to assume that in order to verify the chunks I need to have them in order? So, if I fetch lots of chunks in parallel, and I have gaps, then I need to fill out the gaps before I can verify those because the proofs do not work individually but only in conjunction with the neighbor. Or am I totally off with that?

No, chunks can be delivered in any order, because each of them have a proof

> this would be something entirely new that needs to be adopted

Yes, correct, but if this format turns out to be superior to both others, we can get it adopted reasonably quickly.

> But if your proposed scheme does indeed work and improve the speed of the state sync, I guess it would be worth drafting an EIP for that :)

Ok, lets do a quick PoC and then EIP

**Comment by cburgdorf** [comment]
>No, chunks can be delivered in any order, because each of them have a proof

Oh ok. Good to hear!

>First, you flatten the state as list of pairs (account;account_data) sorted by the hash of the account address (this is how it appears in the Patricia tree)

Can you elaborate on that for me? Am I right to assume the ordering is to ensure that all clients produce chunks in the same deterministic manner? Because otherwise every client ends up producing different, incompatible chunks.

>Ok, lets do a quick PoC and then EIP

Exciting! :+1:

**Comment by AlexeyAkhunov** [comment]
Here is a little doc I have put together: https://github.com/ledgerwatch/eth_state/blob/master/sync/Hybrid_Sync.pdf
